### ISSUE TITLE
Thread.Sleep behaves incorrectly on Linux if system time changes while thread sleeps

### ISSUE THREAD

**Issue by stager13** [dotnet/coreclr#10918]
Steps to reproduce:

Compile the following program:

```c#
using System;
using System.Threading;

namespace SleepTest
{
    class Program
    {
        static void Main(string[] args)
        {
            const int sleepTimeInS = 5;

            for (int i = 0; i < 100000; i++)
            {
                 Console.WriteLine($"Sleeping for {sleepTimeInS} seconds {i}..."

                 Thread.Sleep(sleepTimeInS * 1000);

                 Console.WriteLine("Ok");
            }
        }
    }
}
```

Run it and set system time to some time back (for example 10 minutes)
Thread.Sleep hangs instead of waking up each 5 seconds. I couldn't reproduce this problem on Windows, this code runs fine there. So it is platform-dependent. I also observed similar problem with AutoResetEvent.WaitOne(timeout) call under Linux.

 

OS version: Ubuntu 16.10 x86-64 (also reproducible on Ubuntu MATE on arm (Raspberry PI))
.NET core version: 1.0.1 (also reproducible on 2.0.0-preview1-005448)

**Comment by janvorli** [comment]
I suspect we are using waits with absolute timeouts for synchronization primitives and use realtime clock source (don't set monotonic source explicitly). I remember I had to handle that when I was designing and implementing the event object for CoreRT PAL.

**Comment by gkhanna79** [comment]
CC @kouvel 

@janvorli Should this be for 2.0?

**Comment by janvorli** [comment]
@gkhanna79 it would be good to fix it for 2.0.0

**Comment by gkhanna79** [comment]
Do you want @kouvel to help take a look?

**Comment by janvorli** [comment]
@gkhanna79 I am working on a fix.

**Comment by kouvel** [comment]
Thanks @janvorli, I'm afraid I won't be able to take a look at this any time soon.

**Comment by janvorli** [comment]
Fixed by #11213
